java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedCoProcessFunction.java

```diff
@@ -150,12 +153,52 @@ public void processElement(
         state.addRow(rowTime, row);
     }
 
+    @Override
+    public void processElement2(
+            Row row,
+            KeyedCoProcessFunction<Row, Row, Row, Row>.Context ctx,
+            Collector<Row> collector)
+            throws Exception {
+        if (state.maxTriggeredTimer.value() == null) {
+            return;
+        }
+
+        final Row accumulatorStates = state.getAccumulatorStates();
+        long timestamp = ((Instant) row.getFieldAs(rowTimeFieldName)).toEpochMilli();
+        int lateDataIndex = state.addLateDataAndReference(row);
+
+        for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
+                aggregationFieldsDescriptor.getAggFieldDescriptors()) {
+            final int aggFieldIdx = aggregationFieldsDescriptor.getAggFieldIdx(descriptor);
+            long expireTime =
+                    timestamp
+                            - (timestamp % stepSizeMs)
+                            + descriptor.windowSizeMs
+                            + ((state.maxTriggeredTimer.value() + 1) % stepSizeMs)
+                            - 1
+                            + stepSizeMs;
+            if (state.maxTriggeredTimer.value() >= expireTime) {
+                state.dereferenceLateDataByIndex(lateDataIndex, aggFieldIdx);
+                continue;
+            }
+
+            Object accumulator = accumulatorStates.getField(aggFieldIdx);
+            Object value = row.getField(descriptor.inFieldName);
+            descriptor.aggFunc.add(accumulator, value, timestamp);
```

Comment:
Simplifying the code here might introduce some performance overhead. For example, now we does the following steps:

1. read accumulator value
2. add some values to accumulator
3. retract some values from accumulator
4. save accumulator value

After we simplify them into functions:

1. invoke add_value() function
    a. read accumulator value
    a. add some values to accumulator
    a. save accumulator value
2. invoke retract_value() function
    a. read accumulator value
    a. retract some values from accumulator
    a. save accumulator value

Compared with the previous process, the simplified process read/write the accumulator value from flink state twice. We have not tested against this behavior, not knowing how much overhead it might bring to the performance.